widgetssdk/src/main/java/com/glia/widgets/view/head/controller/ApplicationChatHeadLayoutController.kt

```diff
@@ -0,0 +1,212 @@
+    private var state = State.ENDED
+    private var operatorProfileImgUrl: String? = null
+    private var unreadMessagesCount = 0
+    private var isOnHold = false
+
+    /*
+     * We need to keep track of the currently active (topmost) view. This can be either ChatView
+     * or CallView. CallView has translucent theme and this changes the usual lifecycle of an activity.
+     * When the translucent CallActivity is on top of the ChatActivity and config change happens (e.g screen rotation)
+     * then ChatActivity onResume and onPause are called right after CallActivity's onResume. Current
+     * solution ignores such onResume calls because another activity is already in resumed state.
+     */
+    private var resumedViewName: String? = null
+    override fun onChatHeadClicked() {
+        val destination = navigationDestinationUseCase.execute() ?: return
+        when (destination) {
+            Destinations.CALL_VIEW -> chatHeadLayout?.navigateToCall()
+            Destinations.CHAT_VIEW -> chatHeadLayout?.navigateToChat()
+            Destinations.SCREEN_SHARING -> chatHeadLayout?.navigateToEndScreenSharing()
+        }
+    }
+
+    override fun setView(view: ChatHeadLayoutContract.View) {
+        chatHeadLayout = view
+        init()
+    }
+
+    override fun onDestroy() {
+        if (!isDisplayApplicationChatHeadUseCase(resumedViewName)) { chatHeadLayout?.hide() }
+        gliaOnEngagementUseCase.unregisterListener { newEngagementLoaded() }
+        gliaOnCallVisualizerUseCase.unregisterListener(object :
+            GliaOnCallVisualizerUseCase.Listener {
+            override fun newEngagementLoaded(engagement: OmnibrowseEngagement) {
+                this.newEngagementLoaded(engagement)
+            }
+        })
+        messagesNotSeenHandler.removeListener { count: Int -> onUnreadMessageCountChange(count) }
+        onEngagementEndUseCase.unregisterListener { engagementEnded() }
+        onCallVisualizerEndUseCase.unregisterListener(object :
+            GliaOnCallVisualizerEndUseCase.Listener {
+            override fun engagementEnded() {
+                engagementEnded()
+            }
+        })
+        removeVisitorMediaStateListenerUseCase.execute(this)
+    }
+
+    override fun onNewVisitorMediaState(visitorMediaState: VisitorMediaState?) {
+        // No-op
+    }
+
+    override fun onHoldChanged(isOnHold: Boolean) {
+        this.isOnHold = isOnHold
+    }
+
+    fun updateChatHeadView() {
+        updateChatHeadViewState(chatHeadLayout)
+        updateOnHold()
+        chatHeadLayout?.showUnreadMessageCount(unreadMessagesCount)
+    }
+
+    fun shouldShow(gliaOrRootViewName: String?): Boolean {
+        return isDisplayApplicationChatHeadUseCase(gliaOrRootViewName)
+    }
+
+    private fun init() {
+        gliaOnEngagementUseCase.execute { newEngagementLoaded() }
+        gliaOnCallVisualizerUseCase.execute(object : GliaOnCallVisualizerUseCase.Listener {
```

Comment:
NIT: It is possible to turn `GliaOnCallVisualizerUseCase.Listener` into a `functional interface` and use here as usual lambdas e.g. 

`fun interface Listener {
fun newEngagementLoaded(engagement: OmnibrowseEngagement)
}`
and then
`gliaOnCallVisualizerUseCase.execute {
newEngagementLoaded(it)
}`

Reply:
Thank you for the idea. I've refactored it before your comment a little bit differently, now there is no `object` anymore.